pshtt/pshtt.py

```diff
@@ -526,31 +526,40 @@ def hsts_check(endpoint):
             endpoint.hsts = False
             return
 
-        endpoint.hsts = True
         endpoint.hsts_header = header
 
-        # Set max age to the string after max-age
-        # TODO: make this more resilient to pathological HSTS headers.
-
         # handle multiple HSTS headers, requests comma-separates them
-        first_pass = re.split(r',\s?', header)[0]
-        second_pass = re.sub(r'\'', '', first_pass)
+        headers = [x.strip() for x in header.split(",")]
+        # Multiple HSTS headers does not conform to RFCs 7230-3.2.2 and 6797-6.1
+        if len(headers) > 1:
+            logging.warning("Host is incorrectly returning multiple HSTS headers: {}".format(header))
+
+        # Put all of the directives in the HSTS header into a dictionary
+        directive_list = [x.strip() for x in headers[0].split(";")]
+        directives = dict()
+        for directive in directive_list:
+            components = directive.split("=")
+            directives[components[0]] = "".join(components[1:]) or True
+
+        # max-age is a required directive for HSTS headers
+        if "max-age" not in directives:
+            endpoint.hsts = False
+            return
 
-        temp = re.split(r';\s?', second_pass)
+        endpoint.hsts = True
 
-        if "max-age" in header.lower():
-            endpoint.hsts_max_age = int(temp[0][len("max-age="):])
+        endpoint.hsts_max_age = int(directives["max-age"]) if "max-age" in directives else None
 
         if endpoint.hsts_max_age is None or endpoint.hsts_max_age <= 0:
```

Comment:
Are these values disallowed according to the RFC?  If so we should add a comment stating that.

Reply:
Added a comment explaining what kind of value we're expecting to help explain why we check that.